Nodejs/Product/TestAdapter/TestExecutorWorker.cs

```diff
@@ -0,0 +1,392 @@
+            }
+
+            var port = 0;
+            if (this.runContext.IsBeingDebugged && startedFromVs)
+            {
+                this.DetachDebugger(vsProcessId);
+                // Ensure that --debug-brk or --inspect-brk is the first argument
+                nodeArgs.Insert(0, GetDebugArgs(nodeVersion, out port));
+            }
+
+            // make sure the tests completed is not signalled.
+            this.testsCompleted.Reset();
+
+            this.nodeProcess = ProcessOutput.Run(
+                nodeExePath,
+                nodeArgs,
+                workingDir,
+                env: null,
+                visible: false,
+                redirector: new TestExecutionRedirector(this.ProcessTestRunnerEmit),
+                quoteArgs: false);
+
+            if (this.runContext.IsBeingDebugged && startedFromVs)
+            {
+                this.AttachDebugger(vsProcessId, port, nodeVersion);
+            }
+
+            var serializedObjects = JsonConvert.SerializeObject(testObjects);
+
+            // Send the process the list of tests to run and wait for it to complete
+            this.nodeProcess.WriteInputLine(serializedObjects);
+
+            // for node 8 the process doesn't automatically exit when debugging, so always detach
+            WaitHandle.WaitAny(new[] { this.nodeProcess.WaitHandle, this.testsCompleted });
+            if (this.runContext.IsBeingDebugged && startedFromVs)
+            {
+                this.DetachDebugger(vsProcessId);
+            }
+
+            // Automatically fail tests that haven't been run by this point (failures in before() hooks)
+            foreach (var notRunTest in this.currentTests)
+            {
+                var result = new TestResult(notRunTest)
+                {
+                    Outcome = TestOutcome.Failed
+                };
+
+                if (this.currentResultObject != null)
+                {
+                    result.Messages.Add(new TestResultMessage(TestResultMessage.StandardOutCategory, this.currentResultObject.stdout));
+                    result.Messages.Add(new TestResultMessage(TestResultMessage.StandardErrorCategory, this.currentResultObject.stderr));
+                }
+                this.frameworkHandle.RecordResult(result);
+                this.frameworkHandle.RecordEnd(notRunTest, TestOutcome.Failed);
+            }
+        }
+
+        private void ProcessTestRunnerEmit(string line)
+        {
+            try
+            {
+                var testEvent = JsonConvert.DeserializeObject<TestEvent>(line);
+                // Extract test from list of tests
+                var tests = this.currentTests.Where(n => n.DisplayName == testEvent.title);
+                if (tests.Any())
+                {
+                    switch (testEvent.type)
```

Comment:
>type [](start = 38, length = 4)

Just to confirm, we're ignoring other event types?

Reply:
yes